### PR TITLE
vpp: init at 24.06

### PR DESCRIPTION
**Pull request by maevii**
## Description of changes

VPP is a high-performance userland network stack for routing/switching on general purpose computers.

Supersedes #290529 & #256601, module & tests will be added later in a separate PR (#347797). Basic multithread routing functionality w/ DPDK was tested on an Intel X520-DA2 NIC.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:134]:
```suggestion
    description = "Fast, scalable layer 2-4 multi-platform network stack running in userspace";
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:63]:
```suggestion
      --replace "plugins tools/vppapigen tools/g2 tools/perftool cmake pkg" \
      "plugins tools/vppapigen tools/g2 tools/perftool cmake"
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:88]:
```suggestion
    ++ lib.optional enableDpdk "-DVPP_USE_SYSTEM_DPDK=ON";
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:98]:
```suggestion
    ++ lib.optional enableDpdk dpdk'
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:102]:
```suggestion
    ++ lib.optional enableRdma rdma-core'.dev;
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:36]: It does not work with shared libs?

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:44]: This should preferable fixed in xdp-tools directly.

**Review by markuskowa — CHANGES_REQUESTED** [review]

File: pkgs/by-name/vp/vpp/package.nix
Comment (romner-set) [reply pkgs/by-name/vp/vpp/package.nix:36]: The RDMA plugin is [hardcoded](https://github.com/FDio/vpp/blob/a7e01a83beadc70d94be64c89cf091aaa941d3d8/src/plugins/rdma/CMakeLists.txt#L21-L22) to use static libs, it might change in the future but currently it's necessary to link it statically.

File: pkgs/by-name/vp/vpp/package.nix
Comment (romner-set) [reply pkgs/by-name/vp/vpp/package.nix:44]: To be completely honest, I pretty much blindly copied this part from [the first PR](https://github.com/NixOS/nixpkgs/blob/eaf95021d2e6beab6f3dfab4fa83031f888e0010/pkgs/servers/vpp/default.nix#L34-L39) to get around a `libxdp.a library not found` error, so until now I haven't looked into why it's necessary, but now that I'm looking into I agree. I'll make another PR, then update VPP once it's merged.

Comment (markuskowa) [reply pkgs/by-name/vp/vpp/package.nix:44]: I just checked. That line is not needed anymore. `xdp-tools.lib` does not have any references to `glibc.dev`

File: pkgs/by-name/vp/vpp/package.nix
Comment (markuskowa) [inline pkgs/by-name/vp/vpp/package.nix:127]: One last thing: nested `with`s should not be used. The `with lib.maintainers` can stay, since the package will probably have more than one maintainer in the future.

```suggestion
  meta = {
    description = "Fast, scalable layer 2-4 multi-platform network stack running in user space";
    homepage = "https://s3-docs.fd.io/vpp/${version}/";
    license = [ lib.licenses.asl20 ];
    maintainers = with lib.maintainers; [ romner-set ];
    mainProgram = "vpp";
    platforms = lib.platforms.linux;
```

File: pkgs/by-name/vp/vpp/package.nix
Comment (romner-set) [reply pkgs/by-name/vp/vpp/package.nix:44]: Alright, I'll incorporate that into the PR as well when I get around to making it. For the time being it's still necessary to override `postInstall` to `""` in this package so the static libs aren't removed.

I'll make the PR change xdp-tools' `postInstall` to work similarly to [SDL2's](https://github.com/NixOS/nixpkgs/blob/54cce652cb480258b12f15b2915e7a6c79ac51ae/pkgs/development/libraries/SDL2/default.nix#L185-L189), after which only the `dontDisableStatic` line will be necessary in VPP.

Comment (markuskowa) [reply pkgs/by-name/vp/vpp/package.nix]: I meant: just remove the nuke-refs line here. That is sufficient for now.

There is no need to modify xdp-tools globally. We should not build static libs unless explicitly requested.

Comment (romner-set) [reply pkgs/by-name/vp/vpp/package.nix]: I think it would be beneficial to make xdp-tools build static libs if requested with `dontDisableStatic = true` (like SDL2 does) instead of having to completely overwrite its `postInstall`. Either way, the change in VPP is implemented, so I think it'd be best to leave this topic be for now and have this discussion in the future PR I'll create.

**Review by markuskowa — APPROVED** [review]
LGTM now. Just waiting for the CI to build everything.

**Comment by romner-set** [comment]
@markuskowa Thanks, could you also add the tag for backporting to 24.05? Would be much easier than creating the PR manually.

**Comment by romner-set** [comment]
Nevermind, just realized I can do it myself. Apologies for the ping.

**Pull request merged**
